### ISSUE TITLE
Reload performance is not up to par

### ISSUE THREAD

**Issue by mfrelink** [Hellowlol/HTPC-Manager#484]
### Reporting Issues:

To ensure that a I has enough information to work with please include all of the following information.

**Use proper markdown syntax to structure your post (i.e. code/log in code blocks).**

**Make sure you provide the following information below:**
- [x] Version
  f8c499a78557ea85566a40fad097854d6a92a176
- [x] Branch
  master2
- [x] Commit hash
  f8c499a78557ea85566a40fad097854d6a92a176
- [x] Operating system
  Linux openflixr 4.4.0-31-generic #50-Ubuntu SMP Wed Jul 13 00:07:12 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
- [x] Python version
  Python 2.7.12 (default, Jul  1 2016, 15:12:24)
- [x] What you did?
  Refreshing page ;)
- [x] What happened?
  It is slow sometimes
- [x] What you expected?
  That it would be fast
- [x] How can we reproduce your issue?
  I can't, it's erratic
- [x] What are your (relevant) settings?
  You tell me :) I can give you any info you want but there aren't any settings I think are relevant.
- [x] Include a link to your **FULL** (not just a few lines!) log file that has the error. Please use [Gist](http://gist.github.com) or [Pastebin](http://pastebin.com/).
  There is no error log, only browser performance monitoring

Close your issue when it's solved! If you found the solution yourself please comment so that others benefit from it.

So to be clear.
Somehow, I can't tell you exactly when or in what situation yet, sometimes HTPCmanager is real slow to load.
It takes 5 seconds for the frontage the load.
This only happens (if it happens) when I reload the page, not when accessing menu items or click on the logo in the top left corner.
I discovered the culprit is /js/libs/jquery.treegrid.bootstrap2.js

Why is that, can it be solved?

**Comment by Hellowlol** [comment]
Well it cant be because the filesize as its really small. 

Is authentication used?
What loads slow? The dashboard or the login page?
 is the cnd option checked?

**Comment by mfrelink** [comment]
No authentication is used. So it's the dashboard that loads slow.
It could be related to having system info on the dashboard (or drive info) but I'm not sure yet. 

I played with the CDN option but I don't believe it makes a difference.

**Comment by Hellowlol** [comment]
How many modules does you have active on the dashboard? 

Can you post a screenshot of chrome devtools/firebug on that file, i would like to see why its slow, if its waiting, painting or what ever it slow for :P (simple reload the page again with dev tools open.)

**Comment by mfrelink** [comment]
Update, I have been troubleshooting this for the past hour and it's clear it's not a HTPC Manager problem. Speeds are 100% perfect when I go directly to ip:8085/htpc.

But, as you've guessed from above, I reverse proxy it.
Using nginx with basic settings. I can't seem to solve it though and I have reverse proxied a LOT of apps in this setup and they all work perfectly fine except for HTPC Manager.

If and when I find the solution, I will let you know :)

Thanks for the help!

**Comment by WzL** [comment]
I also use nginx as a reverse proxy icm HTPC Manager and currently have no
performance issues what so ever, so it can work.

**Comment by Hellowlol** [comment]
wzl, wold you mind sharing your relevant config (just mask domain etc)

**Comment by mfrelink** [comment]
For the record, this is what I use:

```
location /htpc {
    proxy_pass http://127.0.0.1:8085;
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        }
```

**Comment by WzL** [comment]
This is what i use, i commented out the proxy headers for some reason which i forgot ;-S but currenlty it works fine for me.

```
location /htpc {
    proxy_pass http://127.0.0.1:8085;
    proxy_set_header Host $host;
#    proxy_set_header X-Real-IP $remote_addr;
#    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
}
```

Settings within HTPC Manager:
Webdir = /htpc
Use rev proxy headers = no
Rev proxy headers base path = 
Use secure headers = yes

**Comment by mfrelink** [comment]
I have exactly the same setup. 
Must be some kind of nginx setting elsewhere I guess.

Still troubleshooting, it's not that big of a deal but it's the last thing for a perfect fast setup so I -will- find it ;)

**Comment by WzL** [comment]
Which OS and version are you running? Are you using software for firewalling purposes on the same server?

**Comment by mfrelink** [comment]
Ubuntu 16.04. I don't use any firewalling software like mod security.
I can't wrap my head around it. Yesterday I was only seeing /js/libs/jquery.treegrid.bootstrap2.js that would stall, now I see around 12 .css and .js files that have TTFB of ~5 seconds and all of them are what you would expect 304, not modified. 

Nothing happens when I disable proxy buffering, same problem.

**Comment by WzL** [comment]
if you change within HTPC Settings the setting LogLevel to Error. Are there any relevant messages possibly thrown in log? Is your Ubuntu installation up to date and did you follow installation instructions on the HTPC page,if not how did you install HTPC manager?

**Comment by mfrelink** [comment]
No errors :) I'm going to do some nginx application logging now to see where the problem is.
Yeah system is fully up-to-date. HTPC Manager is from source. Everything works perfectly fine, I have no issues whatsoever except when I reverse proxy HTPC Manager and reload the page in a browser instead of following links there is a 5 second delay.

You would normally do this once so It's not a big issue, but I want perfection.

**Comment by WzL** [comment]
understood, how about name resolution? Btw, perfection often comes at high costs ;-)

**Comment by mfrelink** [comment]
I already had ruled out DNS, it was also one of the first things I thought of but no problems there.

So far I have this. It's really strange. There's a 5 second gap as you can see, everytime the same place. I have already disabled every dashboard module to test but same thing. 

So for test I renamed fullcalendar.min.js but I had to restart HTPC Manager for it to pick up.
After restart everything was super fast, the same as when I went directly to ip:8085/htpc.

To double check I reversed everything to see if the problem came back. Answer: no.
I really have no clue why this happens. For troubleshooting I already restarted nginx a gazillion times but that did not have any effect. Why does restarting HTPC Manager solves this.

```
10.0.1.41 - openflixr [28/Jul/2016:14:05:23 +0200] "GET /htpc/css/fonts/rg-icons.ttf?oc6klh HTTP/2.0" 304 209 0.004 "https://www.supersecretdomainname.com/htpc/css/rg-icons.min.css?1469707518.34" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.103 Safari/537.36"
10.0.1.41 - openflixr [28/Jul/2016:14:05:28 +0200] "GET /htpc/js/libs/fullcalendar.min.js HTTP/2.0" 304 209 10.004 "https://www.supersecretdomainname.com/htpc/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.103 Safari/537.36"
```

**Comment by Hellowlol** [comment]
I have no idea mate :/

**Comment by WzL** [comment]
Me neither :-/

Op 28 jul. 2016 2:50 p.m. schreef "Hellowlol" notifications@github.com:

**Comment by mfrelink** [comment]
30 minutes later, exactly the same problem, different file. 
Restarting HTPC Manager, problem goes away. 
I'm also at a loss :)

**Comment by Hellowlol** [comment]
Hmm, does the problem still go away if you use htpc manager directly? How many concurrent http calls are used when you load the web page?

Try changing this line https://github.com/Hellowlol/HTPC-Manager/blob/master2/htpc/server.py#L35 
to 30

**Comment by mfrelink** [comment]
That's the whole point, I do not have these problems going directly to ip:8085/htpc :)
So it must be something between nginx reverse proxy and HTPC Manager.

I believe Chrome uses 6. In total there are 61 requests made.

**Comment by WzL** [comment]
now i do recall something, that i also initially had an issue in regards to open sessions from HTPC Manager to Nginx. When HTPC manager is slow can you execute the command: netstat -a | grep 8085 and reply with how many lines are then printed on the screen?

**Comment by mfrelink** [comment]
Around 60. But the number of TIME_WAIT socket connections doesn't seem to be the problem.
I reloaded HTPC Manager in quick succession after which I had like a 1000 connections but it was still super fast. But after a while, 30 minutes or so, it's slow to load (but not directly going to it) even when there are just 2 TIME_WAIT socket connections.

**Comment by WzL** [comment]
you could do some keepalive tuning as described [here](https://ma.ttias.be/enable-keepalive-connections-in-nginx-upstream-proxy-configurations/) , although, as you said yourself, that may not be the issue. I assume the HTPC manager is running on the same IP to which the Nginx proxy server points to? And hosts file is also configured correct? You can also ping your hostname for example?

**Comment by mfrelink** [comment]
Already done keep alive tuning. Also reversed it to see if it helps, but no.
Yes, all on the same server and host file is correct and sure I can ping my hostname :)

I'll leave it for a moment, maybe I'll get a brainwave.

**Comment by mfrelink** [comment]
Update: it turns out to be a browser issue....
I was using Chrome to optimise my server and although I did manage to get a nice performance improvement for everything else, the problem of reloading HTPC Manager stayed.

Until I started testing with Safari. Not only was my tuning 10-fold in Safari, HTPC Manager reloading exhibits no problem whatsoever in this browser.

So yeah, there you have it, I'm an idiot :)

Thanks for all the help though!
